src/workerd/io/container-client.c++

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) 2025 Cloudflare, Inc.
+// Licensed under the Apache 2.0 license found in the LICENSE file or at:
+//     https://opensource.org/licenses/Apache-2.0
+
+#include "container-client.h"
+
+#include <workerd/io/container.capnp.h>
+#include <workerd/io/docker-api.capnp.h>
+#include <workerd/jsg/jsg.h>
+
+#include <capnp/compat/json.h>
+#include <capnp/message.h>
+#include <kj/async.h>
+#include <kj/compat/http.h>
+#include <kj/debug.h>
+#include <kj/encoding.h>
+#include <kj/exception.h>
+
+namespace workerd::io {
+
+template <typename T>
+typename T::Builder decodeJsonResponse(kj::String response) {
+  capnp::JsonCodec codec;
+  codec.handleByAnnotation<T>();
+  capnp::MallocMessageBuilder message;
+  auto jsonRoot = message.initRoot<T>();
+  codec.decode(response, jsonRoot);
+  return jsonRoot;
+}
+
+ContainerClient::ContainerClient(capnp::ByteStreamFactory& byteStreamFactory,
+    kj::Timer& timer,
+    kj::Own<kj::HttpClient> network,
+    kj::Own<kj::HttpClient> httpClient,
+    kj::String containerName,
+    kj::String imageName)
+    : byteStreamFactory(byteStreamFactory),
+      timer(timer),
+      network(kj::mv(network)),
+      httpClient(kj::mv(httpClient)),
+      containerName(kj::mv(containerName)),
+      imageName(kj::mv(imageName)) {}
+
+// Docker-specific Port implementation that implements rpc::Container::Port::Server
+class ContainerClient::DockerPort final: public rpc::Container::Port::Server {
+ public:
+  DockerPort(ContainerClient& containerClient, kj::String containerHost, uint16_t containerPort)
+      : containerClient(containerClient),
+        containerHost(kj::mv(containerHost)),
+        containerPort(containerPort) {}
+
+  kj::Promise<void> connect(ConnectContext context) override {
+    kj::HttpHeaderTable headerTable;
+    kj::HttpHeaders headers(headerTable);
+
+    // Port mappings might have outdated mappings, we can't know if a connect request
+    // fails because the app hasn't finished starting up or because the mapping is outdated.
+    // To be safe we should inspect the container to get up to date mappings.
+    auto portMappings = kj::get<1>(co_await containerClient.inspectContainer());
+    auto maybeMappedPort = portMappings.find(containerPort);
+    if (maybeMappedPort == kj::none) {
+      throw KJ_EXCEPTION(DISCONNECTED, "connect(): Connection refused: container port not found");
+    }
+    auto mappedPort = KJ_ASSERT_NONNULL(maybeMappedPort);
+
+    auto request = containerClient.network->connect(
+        kj::str(containerHost, ":", mappedPort), headers, kj::HttpConnectSettings());
+
+    auto status = co_await request.status;
+    if (status.statusCode != 200) {
+      throw KJ_EXCEPTION(DISCONNECTED, "connect(): Connection refused");
```

Comment:
Ditto... `FAILED` would be better here than `DISCONNECTED` I think.

Reply:
It needs to be DISCONNECTED for it to match what connect is expecting.